File: services/question_service.py
from models.question import Question
from extensions import db
from sqlalchemy import or_

def add_question(data):
    """
    添加题目
    """
    question = Question(
        paper_id=data.get('paper_id'),
        content=data.get('content'),
        type=data.get('type'),
        options=data.get('options'),
        answer=data.get('answer'),
        explanation=data.get('explanation'),
        image_path=data.get('image_path')
    )
    db.session.add(question)
    db.session.commit()
    return question

def update_question_image(question_id, image_path):
    """
    更新题目图片路径
    """
    question = Question.query.get(question_id)
    if question:
        question.image_path = image_path
        db.session.commit()
        return question
    return None

def get_questions_by_paper(paper_id):
    """
    根据试卷ID获取题目列表
    如果没有数据返回None
    """
    questions = Question.query.filter_by(paper_id=paper_id).order_by(Question.id.asc()).all()
    return questions if questions else None

def get_all_questions():
    """
    获取所有题目
    如果没有数据返回None
    """
    questions = Question.query.order_by(Question.created_at.desc()).all()
    return questions if questions else None

def search_questions(keyword=None, paper_id=None, qtype=None, page=1, per_page=10):
    query = Question.query
    if keyword:
        query = query.filter(
            or_(
                Question.content.like(f'%{keyword}%'),
                Question.options.like(f'%{keyword}%'),
                Question.answer.like(f'%{keyword}%'),
                Question.explanation.like(f'%{keyword}%')
            )
        )
    if paper_id:
        query = query.filter_by(paper_id=paper_id)
    if qtype:
        query = query.filter_by(type=qtype)
    pagination = query.order_by(Question.created_at.desc()).paginate(page=page, per_page=per_page, error_out=False)
    return pagination 

File: backend-go/services/question_service.go
package services

import (
	"backend-go/database"
	"backend-go/models"
)

type QuestionInput struct {
	PaperID     uint   `json:"paper_id"`
	Content     string `json:"content"`
	Type        string `json:"type"`
	Options     string `json:"options"`
	Answer      string `json:"answer"`
	Explanation string `json:"explanation"`
	ImagePath   string `json:"image_path"`
}

func AddQuestion(input QuestionInput) (*models.Question, error) {
	question := models.Question{
		PaperID:     input.PaperID,
		Content:     input.Content,
		Type:        input.Type,
		Options:     input.Options,
		Answer:      input.Answer,
		Explanation: input.Explanation,
		ImagePath:   input.ImagePath,
	}

	if err := database.DB.Create(&question).Error; err != nil {
		return nil, err
	}

	return &question, nil
}

func UpdateQuestionImage(questionID uint, imagePath string) (*models.Question, error) {
	var question models.Question
	if err := database.DB.First(&question, questionID).Error; err != nil {
		return nil, err
	}

	question.ImagePath = imagePath
	if err := database.DB.Save(&question).Error; err != nil {
		return nil, err
	}

	return &question, nil
}

func GetQuestionsByPaper(paperID uint) ([]models.Question, error) {
	var questions []models.Question
	if err := database.DB.Where("paper_id = ?", paperID).Order("id ASC").Find(&questions).Error; err != nil {
		return nil, err
	}
	return questions, nil
}

func GetAllQuestions() ([]models.Question, error) {
	var questions []models.Question
	if err := database.DB.Order("created_at DESC").Find(&questions).Error; err != nil {
		return nil, err
	}
	return questions, nil
}

func SearchQuestions(keyword string, paperID uint, qtype string, page, perPage int) ([]models.Question, int64, error) {
	var questions []models.Question
	var total int64

	query := database.DB.Model(&models.Question{})

	if keyword != "" {
		query = query.Where("content LIKE ? OR options LIKE ? OR answer LIKE ? OR explanation LIKE ?",
			"%"+keyword+"%", "%"+keyword+"%", "%"+keyword+"%", "%"+keyword+"%")
	}
	if paperID > 0 {
		query = query.Where("paper_id = ?", paperID)
	}
	if qtype != "" {
		query = query.Where("type = ?", qtype)
	}

	query.Count(&total)

	offset := (page - 1) * perPage
	if err := query.Order("created_at DESC").Offset(offset).Limit(perPage).Find(&questions).Error; err != nil {
		return nil, 0, err
	}

	return questions, total, nil
} 